### DISCUSSION TITLE
Credentials refreshed with every api request

### DISCUSSION THREAD

**Discussion by mikescandy** [aws/aws-sdk-net#3096] [Q&A] [answered]
While memory profiling our webapp, I noticed that the sdk seems to refresh and validate the configuration on each request
![image](https://github.com/aws/aws-sdk-net/assets/2494505/ba72c80c-7cf1-42ed-ab68-ff748e2e35d9)

The screenshot is taken from memory allocations captured during a web requests that makes a dynamodb call. 
The dynamodb client is registered as a singleton, and I made a few requests before this one.

This produces a sensible amount of allocations so it capture my curiosity.

I would have assumed that these credentials would be cached or loaded only once, so now I'm wondering if this is the expected behavior or I'm doing something wrong in the application configuration

**Comment by ashishdhingra** [comment]
@mikescandy Just out of curiosity, are you using DynamoDB local or non-AWS service like simulation (like localstack) by any chance locally for testing the above code for memory profiling? Looking at the logic for [AWS4Signer.DetermineSigningRegion()](https://github.com/aws/aws-sdk-net/blob/1e799114083ecc98549bf96dd376887d5c0557c2/sdk/src/Core/Amazon.Runtime/Internal/Auth/AWS4Signer.cs#L666), there is a code comment which states that `We always have request.AuthenticationRegion defined, as per Amazon.Runtime.Internal.BaseEndpointResolver implementation. request.AuthenticationRegion value is set either based on endpoint rules or overridden by clientConfig.AuthenticationRegion if defined. Normally, users should only override clientConfig.AuthenticationRegion value for non-AWS services`. If `AuthenticationRegion` is somehow not found, it inspects `ClientConfig` for [ServiceURL](https://github.com/aws/aws-sdk-net/blob/1e799114083ecc98549bf96dd376887d5c0557c2/sdk/src/Core/Amazon.Runtime/ClientConfig.cs#L255), which could inspect shared credentials file in certain scenarios. While inspecting shared credentials file, it invokes `Refresh` (as noted in the screenshot in issue description), before processing configured profiles.

Thanks,
Ashish

**Comment by normj** [comment, marked as answer]
Are you creating the service client for every request? The `ServiceUrl` property in your call stack caches the result of the property for subsequent calls. But if you are creating new service clients for every request the cache results will be lost.

https://github.com/aws/aws-sdk-net/blob/main/sdk/src/Core/Amazon.Runtime/ClientConfig.cs#L259

**Comment by mikescandy** [comment]
Thanks for the quick reply!
Everything is registered as a singleton, so it should be reused.
I'll try to isolate some pieces of configuration I can share here to check if I'm doing anything wrong
